rbdmap: replace rbdmap with systemd unit generator (#40844)
ceph/ceph: doc/man/8/rbdmap-generator.rst

boot.

Empty lines and lines starting with the "#" character are ignored. Each of the
remaining lines describes one encrypted block device. Fields are delimited by
white space.

Each line is in the form::

    image-or-snap-spec options

#. image-or-snap-spec is the image or snapshot specification, in the form of
[<pool-name>/[<namespace>/]]<image-name>[@<snap-name>]. If only <image-name> is
specified, <pool-name> defaults to "rbd".

#. options is a comma-delimited list of options. This field is optional. The
supported options are listed below.

If successful, the generated unit maps the image to a ``/dev/rbdX`` device, at
which point a udev rule is triggered to create a friendly device name symlink,
``/dev/rbd/<image-or-snap-spec>``, pointing to the real mapped device.

When writing ``/etc/fstab`` entries for RBD images, The _netdev mount option
should be used. Otherwise, a dependency loop might be created where the mount
point will be pulled in by local-fs.target, while the service to configure the
network is usually only started after the local file system has been mounted.

Supported Options in Config File
--------------------------------

> .. option:: conf=, cluster=, id=, name=, mon_host=, keyfile=, keyring=, device-type=, read-only, exclusive, quiesce=, quiesce-hook=

Author: This line is too long and formatted badly on read-the-docs. Do you have any idea how can I fix this?

https://ceph--40844.org.readthedocs.build/en/40844/man/8/rbdmap-generator/#cmdoption-rbdmap-generator-arg-conf

Reviewer: could just warp this line, like
```rst
.. option:: conf=, cluster=, id=, name=, mon_host=, keyfile=, keyring=,
   device-type=, read-only, exclusive, quiesce=, quiesce-hook=
```

Author: Unfortunately, this makes the command-line man page ugly
![图片](https://user-images.githubusercontent.com/15710934/114701006-d74ea000-9d54-11eb-9636-004baf0adc76.png)

Reviewer: indeed. how about using multiple `.. option::` directives?

>    These options are passed to :doc:`rbd <rbd>`\(8) command directly as

Reviewer: ```suggestion
   These options are passed to :program:`rbd`\(8) command directly as
```

as `rbd` is also defined as under the "program" domain, and can be referenced by `:program:` directly.

   command-line arguments.

.. option:: noauto, auto

>    With noauto, the device will not be added as a dependency for rbdmap.target.

Reviewer: might want to quote `noauto` with backquotes like
```rst
``noauto``
```

   This means that it will not be automatically mapped on boot, unless something
   else pulls it in. In particular, if the device is used for a mount point,

>    it'll be mapped automatically during boot, unless the mount point itself is

Reviewer: ```suggestion
   it is mapped automatically during boot, unless the mount point itself is
```

Author: This paragraph is mostly copied from https://www.freedesktop.org/software/systemd/man/crypttab.html#noauto. And I think the grammar is OK here.

Reviewer: i don't think copying from an existing document could prove the correctness or readability of the resulting document. but i agree the grammar here is correct. what i am not sure is that "will" should be generally used in this context.

@zdover23 what do you think?

   mount point, it'll be mapped automatically during boot, unless the mount
   point itself is also disabled with ``noauto``.

.. option:: nofail, fail

   With ``nofail``, this device will not be a hard dependency of rbdmap.target.
   It'll still be pulled in and started, but the system will not wait for the
   device to show up, and boot will not fail if this is unsuccessful. Note that
   other units that depend on the RBD device may still fail. In particular, if
   the device is used for a mount point, the mount point itself also needs to
   have the ``nofail`` option, or the boot will fail if the device is not mapped
   successfully.

.. option:: x-systemd.requires=, x-systemd.requires-mounts-for=, x-systemd.before=, \
   x-systemd.after=, x-systemd.wanted-by=, x-systemd.required-by=

   See systemd.unit(5) for details.

All other options that not start with "x-" are passed to ``rbd device map``
command directly as command-line arguments.. See :program:`rbd`\(8) for details

Examples
========

Example ``/etc/ceph/rbdtab`` for three RBD images: "bar1" in pool "rbd"; "bar2"
and "bar3" in pool "foopool"::

    krbd bar1
    krbd foopool/bar2    id=admin,noauto

>     krbd foopool/bar3    id=admin        lock_on_read,queue_depth=1024  force

Author: Good point. I think maybe we can switch to a more complex config format? e.g. INI style like [systemd.netdev(5)](https://www.freedesktop.org/software/systemd/man/systemd.netdev.html)

`/etc/ceph/rbd/foopool-bar2.rbd`
```ini
[RBD]
Spec=foopool/bar2
Auto=no
Id=admin

[Mount]
LockOnRead=yes
QueueDepth=1024

[Umount]
force=yes
```

>     krbd foopool/bar3    id=admin        lock_on_read,queue_depth=1024  force

Author: crypttab uses "none" or "-" as a placeholder for its third field, we can also follow it. So that we can write `krbd foopool/bar3 none lock_on_read`. What do you think?

>     krbd foopool/bar3    id=admin        lock_on_read,queue_depth=1024  force

Reviewer 2: LGTM
